system_metrics_collector/test/system_metrics_collector/test_linux_process_cpu_measurement_node.cpp

```diff
@@ -0,0 +1,241 @@
+using moving_average_statistics::StatisticData;
+
+namespace
+{
+constexpr const char kTestNodeName[] = "test_measure_linux_process_cpu";
+constexpr const char kTestTopic[] = "test_process_cpu_measure_topic";
+}
+
+class TestLinuxProcessCpuMeasurementNode : public system_metrics_collector::
+  LinuxProcessCpuMeasurementNode
+{
+public:
+  TestLinuxProcessCpuMeasurementNode(
+    const std::string & name,
+    const std::chrono::milliseconds measurement_period,
+    const std::string & topic,
+    const std::chrono::milliseconds publish_period)
+  : LinuxProcessCpuMeasurementNode(name, measurement_period, topic, publish_period) {}
+
+  /**
+   * Exposes the protected member function for testing purposes.
+   * See description for LinuxProcessCpuMeasurementNode::PeriodicMeasurement().
+   *
+   * @return percentage of CPU this process used
+   */
+  double PeriodicMeasurement() override
+  {
+    LinuxProcessCpuMeasurementNode::PeriodicMeasurement();
+  }
+
+  /**
+   * Exposes the protected member function for testing purposes.
+   * See description for LinuxProcessCpuMeasurementNode::GetMetricName().
+   *
+   * @return a string of the name for this measured metric
+   */
+  std::string GetMetricName() const override
+  {
+    return LinuxProcessCpuMeasurementNode::GetMetricName();
+  }
+};
+
+class TestReceiveProcessCpuMeasurementNode : public rclcpp::Node
+{
+public:
+  explicit TestReceiveProcessCpuMeasurementNode(
+    const std::string & name,
+    const std::string & metric)
+  : rclcpp::Node(name), expected_metric_name_(metric), times_received_(0)
+  {
+    auto callback = [this](MetricsMessage::UniquePtr msg) {this->MetricsMessageCallback(*msg);};
+    subscription_ = create_subscription<MetricsMessage,
+        std::function<void(MetricsMessage::UniquePtr)>>(kTestTopic, 10 /*history_depth*/, callback);
+  }
+
+  int GetNumReceived() const
+  {
+    return times_received_;
+  }
+
+private:
+  void MetricsMessageCallback(const MetricsMessage & msg) const
+  {
+    // Given kPublishPeriod is 80 ms and kTestDuration is 250 ms, the expectation is:
+    // MetricsMessages are published/received at 80 ms, 160 ms, and 240 ms during the first round.
+    // The TestLinuxProcessCpuMeasurementNode is then stopped and restarted and again
+    // MetricsMessages are published/received at 80 ms, 160 ms, and 240 ms during the second round.
+    // This means that no more than 6 MetricsMessages are received.
+    ASSERT_GT(6, times_received_);
+
+    // check source names
+    EXPECT_EQ(kTestNodeName, msg.measurement_source_name);
+    EXPECT_EQ(expected_metric_name_, msg.metrics_source);
+
+    // Check measurements.
+    // There are five types of statistics:
+    // average, maximum, minimum, standard deviation, number of samples
+    EXPECT_EQ(5, msg.statistics.size());
+
+    for (const StatisticDataPoint & stat : msg.statistics) {
+      EXPECT_GT(stat.data_type, 0);
+      if (!std::isnan(stat.data)) {
```

Comment:
Is this check b/c the data hasn't been initialized yet?

Reply:
This applies to CPU measurement nodes, because calculating CPU usage requires 2 measurements, at the beginning and end of a measurement window. The first statistic message that's published has not yet seen the end of the first measurement window, so the statistic ends up being NaN.

(Memory usage measurements is not affected by this problem.)

Reply:
In real world cases, the frequency of making measurements should be significantly higher than the frequency of publishing statistic messages, so NaN statistics would not be seen, but for testing I purposely made their frequencies on the same order of magnitude.